File: scaffold/templates.go
package scaffold

import "fmt"

func mirrorstackYAML(name, title string) string {
	return fmt.Sprintf(`id: %s
name: %s
description: ""
icon: extension
category: content
version: "0.1.0"

dependencies: []
optional_dependencies: []

platform:
  nav_items:
    - icon: extension
      label: %s
      route: /%s

  pages:
    - route: /%s
      component: %sPage

app:
  pages:
    - route: /%s
      component: %sPage
`, name, title, title, name, name, title, name, title)
}

func mirrorstackMD(name, title string) string {
	return fmt.Sprintf(`# %s Module

TODO: Describe what this module does.

## Capabilities
- TODO

## When to use
- TODO

## Data model
- TODO

## Relationships
- Depends on: TODO
`, title)
}

func gitignore() string {
	return `.DS_Store
api/bootstrap
api/lambda.zip
web/dist/
web/node_modules/
`
}

func initialUpSQL() string {
	return `-- Create your tables here
-- This runs inside each app's schema (search_path is set automatically)

-- CREATE TABLE items (
--     id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
--     title TEXT NOT NULL,
--     created_at TIMESTAMPTZ NOT NULL DEFAULT now()
-- );
`
}

func initialDownSQL() string {
	return `-- Reverse of 0000_initial.up.sql

-- DROP TABLE IF EXISTS items;
`
}

func goMod(name string) string {
	return fmt.Sprintf(`module github.com/mirrorstack-ai/app-mod-%s

go 1.24

// require github.com/mirrorstack-ai/app-module-sdk v0.1.0
`, name)
}

func moduleGo(name, title string) string {
	return fmt.Sprintf(`package %s

import (
	"github.com/go-chi/chi/v5"
	// modulesdk "github.com/mirrorstack-ai/app-module-sdk"
)

type Module struct {
	// pool *pgxpool.Pool
}

func New() *Module {
	return &Module{}
}

func (m *Module) ID() string {
	return "%s"
}

func (m *Module) Routes(r chi.Router) {
	// TODO: mount handlers
}
`, name, name)
}

func cmdMainGo(name string) string {
	return fmt.Sprintf(`package main

import (
	"fmt"
	"net/http"
	"os"

	"github.com/go-chi/chi/v5"
	// modulesdk "github.com/mirrorstack-ai/app-module-sdk"
	// "%s" module package
)

func main() {
	r := chi.NewRouter()
	// r.Use(modulesdk.ExtractContext)

	r.Get("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("ok"))
	})

	// mod := %s.New()
	// mod.Routes(r)

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}

	fmt.Printf("listening on :%%s\n", port)
	http.ListenAndServe(":"+port, r)
}
`, name, name)
}

func handlerPublicGo(name string) string {
	return fmt.Sprintf(`package handler

// Public handlers — end user routes (requires app user JWT)

type Handler struct {
	// service *service.%sService
}

func New() *Handler {
	return &Handler{}
}
`, toTitle(name))
}

func handlerAdminGo(name string) string {
	return `package handler

// Admin handlers — app owner routes (requires app user JWT + admin role)
`
}

func serviceGo(name string) string {
	return fmt.Sprintf(`package service

type %sService struct {
	// pool *pgxpool.Pool
}

func New() *%sService {
	return &%sService{}
}
`, toTitle(name), toTitle(name), toTitle(name))
}

func sqlcYAML(name string) string {
	return fmt.Sprintf(`version: "2"
sql:
  - engine: "postgresql"
    queries: "queries/"
    schema: "../../sql/*.up.sql"
    gen:
      go:
        package: "generated"
        out: "generated"
        sql_package: "pgx/v5"
        emit_json_tags: true
        overrides:
          - db_type: "uuid"
            go_type: "github.com/jackc/pgx/v5/pgtype.UUID"
          - db_type: "timestamptz"
            go_type: "github.com/jackc/pgx/v5/pgtype.Timestamptz"
`)
}

func queriesSQL(name string) string {
	return fmt.Sprintf(`-- name: List%s :many
-- SELECT * FROM items ORDER BY created_at DESC LIMIT $1 OFFSET $2;
`, toTitle(name))
}

func webPackageJSON(name string) string {
	return fmt.Sprintf(`{
  "name": "@mirrorstack-ai/mod-%s-web",
  "version": "0.1.0",
  "private": true,
  "type": "module"
}
`, name)
}

func webPlatformIndex(title string) string {
	return fmt.Sprintf(`export { %sPage } from "./pages/%sPage";
`, title, title)
}

func webPlatformPage(title string) string {
	return fmt.Sprintf(`export function %sPage() {
  return (
    <div>
      <h1>%s</h1>
      <p>Platform dashboard page</p>
    </div>
  );
}
`, title, title)
}

func webAppIndex(title string) string {
	return fmt.Sprintf(`export { %sPage } from "./pages/%sPage";
`, title, title)
}

func webAppPage(title string) string {
	return fmt.Sprintf(`export function %sPage() {
  return (
    <div>
      <h1>%s</h1>
      <p>App page</p>
    </div>
  );
}
`, title, title)
}

File: src/scaffold/templates.rs
//! Static templates for the module scaffold. The Go version inlined
//! these as fmt.Sprintf calls; we keep the same shape (raw-string
//! literals + format! for substitutions) so the diff is reviewable
//! against scaffold/templates.go in the Go reference.

/// Returns (relative_path, file_contents) pairs for every file the
/// scaffold creates.
pub fn files(name: &str, title: &str) -> Vec<(String, String)> {
    vec![
        ("mirrorstack.yaml".into(), mirrorstack_yaml(name, title)),
        ("MIRRORSTACK.md".into(), mirrorstack_md(title)),
        (".gitignore".into(), gitignore().into()),
        ("sql/0000_initial.up.sql".into(), initial_up_sql().into()),
        ("sql/0000_initial.down.sql".into(), initial_down_sql().into()),
        ("api/go.mod".into(), go_mod(name)),
        ("api/module.go".into(), module_go(name)),
        ("api/cmd/main.go".into(), cmd_main_go(name)),
        ("api/handler/public.go".into(), handler_public_go(title)),
        ("api/handler/admin.go".into(), handler_admin_go().into()),
        (format!("api/service/{name}.go"), service_go(title)),
        ("api/db/sqlc.yaml".into(), sqlc_yaml().into()),
        (format!("api/db/queries/{name}.sql"), queries_sql(title)),
        ("web/package.json".into(), web_package_json(name)),
        ("web/platform/index.ts".into(), web_index(title)),
        (format!("web/platform/pages/{title}Page.tsx"), web_platform_page(title)),
        ("web/app/index.ts".into(), web_index(title)),
        (format!("web/app/pages/{title}Page.tsx"), web_app_page(title)),
    ]
}

fn mirrorstack_yaml(name: &str, title: &str) -> String {
    format!(
        "id: {name}\n\
         name: {title}\n\
         description: \"\"\n\
         icon: extension\n\
         category: content\n\
         version: \"0.1.0\"\n\
         \n\
         dependencies: []\n\
         optional_dependencies: []\n\
         \n\
         platform:\n\
           nav_items:\n\
             - icon: extension\n\
               label: {title}\n\
               route: /{name}\n\
         \n\
           pages:\n\
             - route: /{name}\n\
               component: {title}Page\n\
         \n\
         app:\n\
           pages:\n\
             - route: /{name}\n\
               component: {title}Page\n",
    )
}

fn mirrorstack_md(title: &str) -> String {
    format!(
        "# {title} Module\n\
         \n\
         TODO: Describe what this module does.\n\
         \n\
         ## Capabilities\n\
         - TODO\n\
         \n\
         ## When to use\n\
         - TODO\n\
         \n\
         ## Data model\n\
         - TODO\n\
         \n\
         ## Relationships\n\
         - Depends on: TODO\n",
    )
}

fn gitignore() -> &'static str {
    ".DS_Store\n\
     api/bootstrap\n\
     api/lambda.zip\n\
     web/dist/\n\
     web/node_modules/\n"
}

fn initial_up_sql() -> &'static str {
    "-- Create your tables here\n\
     -- This runs inside each app's schema (search_path is set automatically)\n\
     \n\
     -- CREATE TABLE items (\n\
     --     id UUID PRIMARY KEY DEFAULT gen_random_uuid(),\n\
     --     title TEXT NOT NULL,\n\
     --     created_at TIMESTAMPTZ NOT NULL DEFAULT now()\n\
     -- );\n"
}

fn initial_down_sql() -> &'static str {
    "-- Reverse of 0000_initial.up.sql\n\
     \n\
     -- DROP TABLE IF EXISTS items;\n"
}

fn go_mod(name: &str) -> String {
    format!(
        "module github.com/mirrorstack-ai/app-mod-{name}\n\
         \n\
         go 1.24\n\
         \n\
         // require github.com/mirrorstack-ai/app-module-sdk v0.1.0\n",
    )
}

fn module_go(name: &str) -> String {
    format!(
        "package {name}\n\
         \n\
         import (\n\
         \t\"github.com/go-chi/chi/v5\"\n\
         \t// modulesdk \"github.com/mirrorstack-ai/app-module-sdk\"\n\
         )\n\
         \n\
         type Module struct {{\n\
         \t// pool *pgxpool.Pool\n\
         }}\n\
         \n\
         func New() *Module {{\n\
         \treturn &Module{{}}\n\
         }}\n\
         \n\
         func (m *Module) ID() string {{\n\
         \treturn \"{name}\"\n\
         }}\n\
         \n\
         func (m *Module) Routes(r chi.Router) {{\n\
         \t// TODO: mount handlers\n\
         }}\n",
    )
}

fn cmd_main_go(name: &str) -> String {
    format!(
        "package main\n\
         \n\
         import (\n\
         \t\"fmt\"\n\
         \t\"net/http\"\n\
         \t\"os\"\n\
         \n\
         \t\"github.com/go-chi/chi/v5\"\n\
         \t// modulesdk \"github.com/mirrorstack-ai/app-module-sdk\"\n\
         \t// \"{name}\" module package\n\
         )\n\
         \n\
         func main() {{\n\
         \tr := chi.NewRouter()\n\
         \t// r.Use(modulesdk.ExtractContext)\n\
         \n\
         \tr.Get(\"/health\", func(w http.ResponseWriter, r *http.Request) {{\n\
         \t\tw.Write([]byte(\"ok\"))\n\
         \t}})\n\
         \n\
         \t// mod := {name}.New()\n\
         \t// mod.Routes(r)\n\
         \n\
         \tport := os.Getenv(\"PORT\")\n\
         \tif port == \"\" {{\n\
         \t\tport = \"8080\"\n\
         \t}}\n\
         \n\
         \tfmt.Printf(\"listening on :%s\\n\", port)\n\
         \thttp.ListenAndServe(\":\"+port, r)\n\
         }}\n",
    )
}

fn handler_public_go(title: &str) -> String {
    format!(
        "package handler\n\
         \n\
         // Public handlers — end user routes (requires app user JWT)\n\
         \n\
         type Handler struct {{\n\
         \t// service *service.{title}Service\n\
         }}\n\
         \n\
         func New() *Handler {{\n\
         \treturn &Handler{{}}\n\
         }}\n",
    )
}

fn handler_admin_go() -> &'static str {
    "package handler\n\
     \n\
     // Admin handlers — app owner routes (requires app user JWT + admin role)\n"
}

fn service_go(title: &str) -> String {
    format!(
        "package service\n\
         \n\
         type {title}Service struct {{\n\
         \t// pool *pgxpool.Pool\n\
         }}\n\
         \n\
         func New() *{title}Service {{\n\
         \treturn &{title}Service{{}}\n\
         }}\n",
    )
}

fn sqlc_yaml() -> &'static str {
    "version: \"2\"\n\
     sql:\n\
       - engine: \"postgresql\"\n\
         queries: \"queries/\"\n\
         schema: \"../../sql/*.up.sql\"\n\
         gen:\n\
           go:\n\
             package: \"generated\"\n\
             out: \"generated\"\n\
             sql_package: \"pgx/v5\"\n\
             emit_json_tags: true\n\
             overrides:\n\
               - db_type: \"uuid\"\n\
                 go_type: \"github.com/jackc/pgx/v5/pgtype.UUID\"\n\
               - db_type: \"timestamptz\"\n\
                 go_type: \"github.com/jackc/pgx/v5/pgtype.Timestamptz\"\n"
}

fn queries_sql(title: &str) -> String {
    format!(
        "-- name: List{title} :many\n\
         -- SELECT * FROM items ORDER BY created_at DESC LIMIT $1 OFFSET $2;\n",
    )
}

fn web_package_json(name: &str) -> String {
    format!(
        "{{\n  \"name\": \"@mirrorstack-ai/mod-{name}-web\",\n  \"version\": \"0.1.0\",\n  \"private\": true,\n  \"type\": \"module\"\n}}\n",
    )
}

fn web_index(title: &str) -> String {
    format!("export {{ {title}Page }} from \"./pages/{title}Page\";\n")
}

fn web_platform_page(title: &str) -> String {
    format!(
        "export function {title}Page() {{\n  return (\n    <div>\n      <h1>{title}</h1>\n      <p>Platform dashboard page</p>\n    </div>\n  );\n}}\n",
    )
}

fn web_app_page(title: &str) -> String {
    format!(
        "export function {title}Page() {{\n  return (\n    <div>\n      <h1>{title}</h1>\n      <p>App page</p>\n    </div>\n  );\n}}\n",
    )
}
